src/include/duckdb/execution/operator/persistent/csv_scanner/csv_state_machine_cache.hpp

```diff
@@ -0,0 +1,30 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// duckdb/execution/operator/persistent/csv_scanner/csv_state_machine_cache.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "duckdb/execution/operator/persistent/csv_scanner/csv_reader_options.hpp"
+#include "duckdb/execution/operator/persistent/csv_scanner/csv_buffer_manager.hpp"
+
+namespace duckdb {
+typedef uint8_t state_machine_t[7][256];
+
+class CSVStateMachineCache {
+public:
+	CSVStateMachineCache();
+	~CSVStateMachineCache() {};
+	//! Gets a state machine from the cache, if it's not from one the default options
+	//! It first caches it, then returns it.
+	state_machine_t &Get(char delimiter, char quote, char escape);
+
+private:
+	void Insert(char delimiter, char quote, char escape);
+	//! Cache on delimiter|quote|escape
+	unordered_map<char, unordered_map<char, unordered_map<char, state_machine_t>>> state_machine_cache;
```

Comment:
I am curious about this. It is a bit hard to read, is there a significant performance reason behind not using a `struct`? If so, then we should add a comment. 

Otherwise, can we not have a struct and add a custom equal operator? Something like this:
```cpp
struct CSVStateMachineConfig {
char delimiter;
char quote;
char escape;
};
```

Reply:
With a struct we have O(N) access no? Not that I expect the cache to get ginormous. I like the O(1) and I don't think is that bad to read

Reply:
`unordered_map<my_struct, state_machine_t>` with a custom comparison function and hash function on the struct should still be `O(1)`, no? I was thinking of something like this, but I could be mistaken. :) It is more extendable (if we get more state machine options, not sure about that). But I can also see that it adds more code—just my thoughts. 😄 

```cpp
struct CSVStateMachineConfig {
	char delimiter;
	char quote;
	char escape;

	bool operator==(const CSVStateMachineConfig &other) const {
	    return delimiter == other.delimiter && quote == other.quote && escape == other.escape;
	}
};

// possibly want to use our hash function (duckdb/common/types/hash.hpp)
struct HashCSVStateMachineConfig {
	std::size_t operator()(CSVStateMachineConfig const& config) const noexcept {
		auto h_delimiter = std::hash<char>{}(config.delimiter);
		auto h_quote = std::hash<char>{}(config.quote);
		auto h_escape = std::hash<char>{}(config.escape);
		return CombineHash(h_delimiter, CombineHash(h_quote, h_escape));
	}
};

unordered_map<CSVStateMachineConfig, state_machine_t> configurations;
```

Reply:
fair enough :-)